projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,280 @@
+
+1) **Enhanced Security**: these devices are specialized to provide a high level of security for cryptographic operations.
+1) **Protection against attacks**: HSM devices help mitigate various cyberattacks by securely storing private keys and performing critical cryptographic operations within a secure environment
+1) **Compliance and regulatory requirements**: in many industries and jurisdictions, compliance with security standards and regulations is crucial. Integrating a certified and recognized HSM device with our wallet can help use cases meet those requirements. Their usage can also demonstrate a commitment to data protection and security best practices.
+1) **Offline key generation**: keys generated within an HSM never come into contact with potentially insecure environments, which adds an extra layer of protection against external threats
+1) **Audit Trail and Accountability**: HSMs often provide robust audit trail capabilities. This feature helps maintain accountability and traceability, crucial in enterprise and compliance-driven industries
+1) **Trust and Confidence**: Integrating an HSM with Hathor instills trust and confidence among use cases. When corporations know that their assets are secured by industry-standard HSMs, they are more likely to trust our platform and use it for their financial and business needs.
+
+Overall, the integration of our wallet with an HSM helps elevate the security posture of your project, reduces the risk of unauthorized access and manipulation, and increases trust among stakeholders.
+
+As the blockchain industry continues to mature, security will remain a paramount concern, making the adoption of HSMs increasingly relevant for safeguarding digital assets and transactions.
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The main premise of the integration with an HSM is that most cryptographic operations should be executed within the secure environment of the hardware device, adding an insulation layer from external threats.
+
+The most important of them, the creation of the _Extended Private Key_ (`xPriv`) is generated offline within the device and never retrieved. All interactions that depend on this `xPriv`, such as signing transactions, are done exclusively through APIs and only the final results are returned.
+
+For the Wallet Lib this means that the following operations are executed within the HSM environment:
+- Generation of a `BIP-32 xPriv`
+- Generation of an `xPub` for this `xPriv`
+- Derivation of all child private keys for this `xPriv`
+- Signing transaction inputs
+
+### The HSM Environment
+The full documentation of the HSM and its API can be found on the [Dinamo HSM docs GitBook](https://docs.dinamonetworks.com/). In short, the architecture of this device from our client standpoint consists of:
+- A web server that provides the API endpoints
+- Each connection to this server consists a **session**, and is authenticated using a username and password
```

Comment:
What happens if the wallet loses connection to this session while using it? Is there any way of listening to this information? Or do we need to send a request to it to check?

Reply:
I think now I understand after reading the code. You never keep the session opened, right? You send the request to the HSM, then disconnects, right? If that's the case, it's good to be clear here in the design (or I might have missed this information)

Reply:
I added some clarifications about the singleton nature of the HSM connections on [62b08e6](https://github.com/HathorNetwork/rfcs/pull/67/commits/62b08e6930f53ccbb681c30ced836f7554354c35).